src/keys/YkChallengeResponseKey.cpp

```diff
@@ -63,60 +57,25 @@ QByteArray YkChallengeResponseKey::rawKey() const
     return QByteArray::fromRawData(m_key, static_cast<int>(m_keySize));
 }
 
-/**
- * Assumes yubikey()->init() was called
- */
-bool YkChallengeResponseKey::challenge(const QByteArray& c)
-{
-    return challenge(c, 2);
-}
-
-bool YkChallengeResponseKey::challenge(const QByteArray& challenge, unsigned int retries)
+bool YkChallengeResponseKey::challenge(const QByteArray& challenge)
 {
-    do {
-        --retries;
-
-        if (m_blocking) {
-            emit userInteractionRequired();
-        }
-
-        QByteArray key;
-        auto result = AsyncTask::runAndWaitForFuture(
-            [this, challenge, &key]() { return YubiKey::instance()->challenge(m_slot, true, challenge, key); });
-
-        if (m_blocking) {
-            emit userConfirmed();
-        }
-
-        if (result == YubiKey::SUCCESS) {
-            if (m_key) {
-                gcry_free(m_key);
-            }
-            m_keySize = static_cast<std::size_t>(key.size());
-            m_key = static_cast<char*>(gcry_malloc_secure(m_keySize));
-            std::memcpy(m_key, key.data(), m_keySize);
-            sodium_memzero(key.data(), static_cast<std::size_t>(key.capacity()));
-            return true;
+    m_error.clear();
+    QByteArray key;
+    auto result =
+        AsyncTask::runAndWaitForFuture([&] { return YubiKey::instance()->challenge(m_keySlot, challenge, key); });
+
+    if (result == YubiKey::SUCCESS) {
+        if (m_key) {
+            gcry_free(m_key);
```

Comment:
I'm not so much a fan of scattered non-RAII frees like this. I prefer we reuse existing (or extended) PasswordKey logic here.

Reply:
This was moved code

Reply:
Then I take the blame for not refactoring that in the first place. :-D

Reply:
I am not a fan of that logic and have seen it in the code. I prefer a properly named container like "SecureBuffer" which would encapsulate this behavior. Then PasswordKey would use that buffer but we don't sprinkle specific use case classes like PasswordKey around the world.